### PR TITLE
Better support for `--llvm-root`.

### PR DESCRIPTION
**Pull request by DiamondLovesYou**
This handles cases when the LLVM used isn't configured will the 'usual' targets. Also, cases where LLVM is shared are also handled (ie with `LD_LIBRARY_PATH` etc).

### CODE REVIEW

File: mk/tests.mk [hunk 1]
```diff
@@ -1064,7 +1064,9 @@ $(3)/test/run-make/%-$(1)-T-$(2)-H-$(3).ok: \
 	    "$$(LD_LIBRARY_PATH_ENV_HOSTDIR$(1)_T_$(2)_H_$(3))" \
 	    "$$(LD_LIBRARY_PATH_ENV_TARGETDIR$(1)_T_$(2)_H_$(3))" \
 	    $(1) \
-	    $$(S)
+	    $$(S) \
+	    "$$(LLVM_LIBDIR_RUSTFLAGS_$(3))" \
+	    "$$(shell $$(LLVM_CONFIG_$(3)) --components)"
```
Comment (alexcrichton) [inline mk/tests.mk]: Can this output be cached in a variable to prevent re-running it?

File: src/librustc_llvm/lib.rs [hunk 2]
```diff
@@ -2327,6 +2328,37 @@ pub unsafe fn debug_loc_to_string(c: ContextRef, tr: DebugLocRef) -> String {
         .expect("got a non-UTF8 DebugLoc from LLVM")
 }
 
+pub fn initialize_available_targets() {
+    macro_rules! init_target(
+        ($cfg:ident $fn_name:ident) => { {
```
Comment (alexcrichton) [inline src/librustc_llvm/lib.rs]: The `fn_name` ident may be better named `arch_name`

File: mk/llvm.mk [hunk 3]
```diff
@@ -83,7 +83,7 @@ endif
 LLVM_LINKAGE_PATH_$(1):=$$(abspath $$(RT_OUTPUT_DIR_$(1))/llvmdeps.rs)
 $$(LLVM_LINKAGE_PATH_$(1)): $(S)src/etc/mklldeps.py $$(LLVM_CONFIG_$(1))
 	$(Q)$(CFG_PYTHON) "$$<" "$$@" "$$(LLVM_COMPONENTS)" "$$(CFG_ENABLE_LLVM_STATIC_STDCPP)" \
-		$$(LLVM_CONFIG_$(1))
+	        "$$(CFG_ENABLE_LLVM_LIBCPP)" $$(LLVM_CONFIG_$(1))
```
Comment (alexcrichton) [inline mk/llvm.mk]: In the configure script it looks like this is `CFG_ENABLE_LIBCPP` instead of `LLVM_LIBCPP`, was this defined elsewhere though?

**Comment by alexcrichton** [comment]
The travis failures here look like they may not be benign

File: mk/llvm.mk (hunk 3 above)
Comment (DiamondLovesYou) [inline mk/llvm.mk]: Actually it's worse than that: `configure` doesn't even `putvar` anything related to `libcpp`. Nice catch, thanks.

File: mk/tests.mk (hunk 1 above)
Comment (DiamondLovesYou) [inline mk/tests.mk]: Done.

File: src/librustc_llvm/lib.rs (hunk 2 above)
Comment (DiamondLovesYou) [inline src/librustc_llvm/lib.rs]: Yeah, silly me. Done.

**Comment by alexcrichton** [comment]
Can this hold off on the changes to tests/maketest for now? The changes to compiletest are only needed if tests are linking to the compiler (e.g. fulldeps), but for most targets with custom LLVM builds we probably don't need the compiler itself working there just yet, so fixes like this can come later if absolutely necessary. Same with the changes to run-make as I believe run-make isn't run for target-only targets (as opposed to host+target targets).

I continue to be very wary of landing all these changes as it's quite a deal of difficult logic that isn't tested in any way, so I prefer to minimize the impact as much as possible.

**Comment by DiamondLovesYou** [comment]
@alexcrichton They were due to variable expansions happening before `llvm-config` was built (ie not an issue if `make all` is run before tidy or check, thus I never caught it locally). Anyway, fixed.

**Comment by brson** [comment]
Looks like good cleanup. I agree that the additional complexity in compiletest is unwarranted for now. Maybe we can fix this problem some other way and I'm happy to ignore all fulldeps tests on the web for a while.

**Comment by DiamondLovesYou** [comment]
@brson I'm not sure I understand what you mean by "I'm happy to ignore all fulldeps tests on the web for a while". Could you explain exactly what you mean?

Also, surely the benefit of the fulldep tests outweighs the pretty minimal additional complexity in `compiletest` (from calling `llvm-config`)? I mean I have no problem just disabling the fulldeps if that's way you'd like to go. I can file an issue and then comment out the fulldep test modes in `tests.mk` (assuming this is what you meant from above).

**Comment by alexcrichton** [comment]
Currently running the fulldeps tests means that the entire compiler has to be built for the target platform, and in this case that probably isn't necessary. I believe @brson means that in the interest of keeping this minimal (e.g. only running the `run-pass` tests for things like PNaCl) the changes to `compiletest` shouldn't be necessary because none of those link to LLVM.

Unfortunately "minimal complexity" ends up being bitrotted very quickly without being exercised which ends up porting it forward for a few years or so only to be painfully deleted at some point.

**Comment by wthrowe** [comment]
Following up on some discussion on #28327: The current patch does not work for my use case, and I don't think the LLVM patch mentioned in the #28327 discussion is going to help, because I don't have a `libLLVM.so` or `libLLVM-${VERSION}.so`, so the current `llvm-config` logic will still apply.  My system has all of the individual LLVM libraries installed as separate shared objects, so `llvm-config` is *correctly* reporting a long list of linker arguments for shared libraries.

**Comment by DiamondLovesYou** [comment]
Is your system's LLVM built with CMake && BUILD_SHARED_LIBS `ON`? CMake
will build separate shared objects for each component in that case.

I'll have to add an extra flag to `llvm-config` to fix that.

**Comment by wthrowe** [comment]
Yes, it looks like it uses CMake with `-DBUILD_SHARED_LIBS=ON`.  (This is Gentoo's official llvm-3.7.0 package.)

**Comment by jauhien** [comment]
Just fyi: we in Gentoo are interested in the possibility to link Rust to separated LLVM shared libraries, otherwise we can't use the system wide LLVM which is quite sad from the point of view of the overall build time.

**Comment by DiamondLovesYou** [comment]
@jauhien I've just created [this](http://reviews.llvm.org/D13198) patch. It fixes `llvm-config` for cmake && `BUILD_SHARED_LIBS`.

**Comment by DiamondLovesYou** [comment]
My patch has landed into LLVM proper and I've removed the extra stuff in `compiletest` w.r.t. fulldeps tests. This is ready to be merged.

File: mk/llvm.mk [hunk 4]
```diff
@@ -95,6 +95,8 @@ $(foreach host,$(CFG_HOST), \
 # This can't be done in target.mk because it's included before this file.
 define LLVM_LINKAGE_DEPS
 $$(TLIB$(1)_T_$(2)_H_$(3))/stamp.rustc_llvm: $$(LLVM_LINKAGE_PATH_$(2))
+RUSTFLAGS$(1)_rustc_llvm_T_$(3) += $$(shell $$(LLVM_CONFIG_$(3)) --components | tr '-' '_' |\
```
Comment (alexcrichton) [inline mk/llvm.mk]: Can this use `$(LLVM_COMPONENTS)` instead of re-invoking llvm-config here?

Comment (alexcrichton) [inline mk/llvm.mk]: This also seems to disagree currently in terms of which llvm-config is run?

File: configure [hunk 5]
```diff
@@ -1732,6 +1738,7 @@ putvar CFG_AARCH64_LINUX_ANDROID_NDK
 putvar CFG_ARM_LINUX_ANDROIDEABI_NDK
 putvar CFG_I686_LINUX_ANDROID_NDK
 putvar CFG_MANDIR
+putvar CFG_USING_LIBCPP
```
Comment (alexcrichton) [inline configure]: Were these changes necessary for now? It looks like they're not needed elsewhere

File: mk/main.mk [hunk 6]
```diff
@@ -476,9 +487,9 @@ endif
 endif
 
 LD_LIBRARY_PATH_ENV_HOSTDIR$(1)_T_$(2)_H_$(3) := \
-    $$(CURDIR)/$$(HLIB$(1)_H_$(3))
+    $$(CURDIR)/$$(HLIB$(1)_H_$(3)):$$(CFG_LLVM_INST_DIR_$(3))/lib
 LD_LIBRARY_PATH_ENV_TARGETDIR$(1)_T_$(2)_H_$(3) := \
-    $$(CURDIR)/$$(TLIB1_T_$(2)_H_$(CFG_BUILD))
+    $$(CURDIR)/$$(TLIB1_T_$(2)_H_$(CFG_BUILD)):$$(CFG_LLVM_INST_DIR_$(CFG_BUILD))/lib
```
Comment (alexcrichton) [inline mk/main.mk]: I don't think that this belongs in the targetdir var because this is only needed for the compiler itself to run

File: src/etc/maketest.py [hunk 7]
```diff
@@ -53,6 +53,8 @@ def convert_path_spec(name, value):
 putenv('TARGET_RPATH_DIR', os.path.abspath(sys.argv[11]))
 putenv('RUST_BUILD_STAGE', sys.argv[12])
 putenv('S', os.path.abspath(sys.argv[13]))
+putenv('RUSTFLAGS', sys.argv[15])
```
Comment (alexcrichton) [inline src/etc/maketest.py]: Was this necessary to pass along? It looks like this isn't used elsewhere from this patch? Also this looks like it skipped a number?

Comment (DiamondLovesYou) [inline src/etc/maketest.py]: > Was this necessary to pass along? It looks like this isn't used elsewhere from this patch?

Yeah, the `execution-engine` test needs access to LLVM's dylib. Since trying to disable that test only when LLVM is shared would also require an additional env var, might as well just fix the test.

> Also this looks like it skipped a number?

Didn't skip a number; `target_triple` from earlier in the file takes arguement 14.

File: configure (hunk 5 above)
Comment (DiamondLovesYou) [inline configure]: They should be needed in `src/etc/mklldeps.py`...

Huh. Somehow, that part of mklldeps didn't make it. I'll get that fixed.

File: mk/llvm.mk (hunk 4 above)
Comment (DiamondLovesYou) [inline mk/llvm.mk]: > Can this use $(LLVM_COMPONENTS) instead of re-invoking llvm-config here?

I don't see why not.

> This also seems to disagree currently in terms of which llvm-config is run?

What do you mean?

File: src/etc/maketest.py (hunk 7 above)
Comment (DiamondLovesYou) [inline src/etc/maketest.py]: Indeed, fixed.

File: mk/main.mk (hunk 6 above)
Comment (DiamondLovesYou) [inline mk/main.mk]: Fixed.

**Comment by DiamondLovesYou** [comment]
Ready to go.

File: mk/llvm.mk [hunk 8]
```diff
@@ -80,10 +80,17 @@ endif
...
+# Note: Filter with llvm-config so that optional targets which aren't present
+# don't cause errors (ie PNaCl's target is only present within PNaCl's LLVM
+# fork).
 LLVM_LINKAGE_PATH_$(1):=$$(abspath $$(RT_OUTPUT_DIR_$(1))/llvmdeps.rs)
 $$(LLVM_LINKAGE_PATH_$(1)): $(S)src/etc/mklldeps.py $$(LLVM_CONFIG_$(1))
-	$(Q)$(CFG_PYTHON) "$$<" "$$@" "$$(LLVM_COMPONENTS)" "$$(CFG_ENABLE_LLVM_STATIC_STDCPP)" \
-		$$(LLVM_CONFIG_$(1)) "$(CFG_STDCPP_NAME)"
+	$(Q)$(CFG_PYTHON) "$$<" "$$@" "$(filter $(shell \
+				$(CFG_LLVM_ROOT)/bin/llvm-config$(X_$(CFG_BUILD)) \
+					--components), \
```
Comment (alexcrichton) [inline mk/llvm.mk]: Can this be replaced with a variable defined below to avoid re-invoking llvm-config?

**Comment by alexcrichton** [comment]
I have a feeling that this is likely to break over time (as there's no CI and this looks pretty brittle as so many build system things are), but it seems fine as it should be easy to get back up to speed in the future.

File: mk/llvm.mk (hunk 8 above)
Comment (DiamondLovesYou) [inline mk/llvm.mk]: No, because we don't want to link unused LLVM components into `rustc`.

Comment (DiamondLovesYou) [inline mk/llvm.mk]: I mean, I don't have a problem with just linking everything, I just thought you guys might not like it.

**Comment by alexcrichton** [comment]
Ok, the point about re-invoking `llvm-config` sounds good to me, can these commits be squashed into one as well? Other than that r=me

**Comment by DiamondLovesYou** [comment]
Done!

**Comment by DiamondLovesYou** [comment]
Whoops, I forgot to push a change to `llvm.mk`.
